sql/core/src/main/scala/org/apache/spark/sql/diagnostic/DiagnosticListener.scala

```diff
@@ -0,0 +1,103 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.diagnostic
+
+import org.apache.spark.SparkConf
+import org.apache.spark.scheduler.{SparkListener, SparkListenerEvent}
+import org.apache.spark.sql.execution.ExplainMode
+import org.apache.spark.sql.execution.ui.{SparkListenerSQLAdaptiveExecutionUpdate, SparkListenerSQLExecutionEnd, SparkListenerSQLExecutionStart}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.internal.StaticSQLConf.UI_RETAINED_EXECUTIONS
+import org.apache.spark.status.{ElementTrackingStore, KVUtils}
+
+class DiagnosticListener(
+    conf: SparkConf,
+    kvStore: ElementTrackingStore) extends SparkListener {
+
+  kvStore.addTrigger(
+    classOf[ExecutionDiagnosticData],
+    conf.get(UI_RETAINED_EXECUTIONS)) { count =>
+    cleanupExecutions(count)
+  }
+
+  override def onOtherEvent(event: SparkListenerEvent): Unit = event match {
+    case e: SparkListenerSQLExecutionStart => onExecutionStart(e)
+    case e: SparkListenerSQLExecutionEnd => onExecutionEnd(e)
+    case e: SparkListenerSQLAdaptiveExecutionUpdate => onAdaptiveExecutionUpdate(e)
+    case _ => // Ignore
+  }
+
+  private def onAdaptiveExecutionUpdate(event: SparkListenerSQLAdaptiveExecutionUpdate): Unit = {
+    val data = new AdaptiveExecutionUpdate(
+      event.executionId,
+      System.currentTimeMillis(),
+      event.physicalPlanDescription
+    )
+    kvStore.write(data)
+  }
+
+  private def onExecutionStart(event: SparkListenerSQLExecutionStart): Unit = {
+    val planDescriptionMode = ExplainMode.fromString(SQLConf.get.uiExplainMode)
+    val physicalPlan = event.qe.explainString(planDescriptionMode, Int.MaxValue)
+    val data = new ExecutionDiagnosticData(
+      event.executionId,
+      physicalPlan,
+      event.time,
+      None,
+      None
+    )
+    // Check triggers since it's adding new netries
+    kvStore.write(data, checkTriggers = true)
+  }
+
+  private def onExecutionEnd(event: SparkListenerSQLExecutionEnd): Unit = {
+    try {
+      val existing = kvStore.read(classOf[ExecutionDiagnosticData], event.executionId)
+      val planDescriptionMode = ExplainMode.fromString(SQLConf.get.uiExplainMode)
+      val physicalPlan = event.qe.explainString(planDescriptionMode, Int.MaxValue)
```

Comment:
Do you know the required disk size with `Int.MaxValue`? It could kill driver pod due to `OutOfDisk`.

Reply:
good point. It might be dangerous to get the untruncated plan description of a very large plan, but it's also hard to estimate the upper bound because theoretically, there is no limitation of the query plan size.
How about we add a flag to control plan truncation for the disk store? for example: `spark.appStatusStore.diskStore.saveUntruncatedPlan=true`

Reply:
I tried very large plans locally with a 2GB memory spark driver. It turns out the Spark itself will OOM far before the query plan string becomes too large (~10MB). In addition, there is a retained number for diagnostic data (1000 by default), so the memory/disk consumption should be fine.

Reply:
Agree with @dongjoon-hyun , we should impose a limit here.

@shardulm94 can comment more on his observations with something similar in terms of increase in cost - he had done a streaming serialization implementation to get around the issue (he was writing to hdfs, so the solution directly wont apply here).

Reply:
done